jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java

```diff
@@ -215,5 +696,1697 @@ public void clear()
+            {
+                for (Iterator<RetainableByteBuffer> i = dynamicCapacity._buffers.iterator(); i.hasNext();)
+                {
+                    RetainableByteBuffer buffer = i.next();
+                    if (!append(buffer))
+                        return false;
+                    buffer.release();
+                    i.remove();
+                }
+                return true;
+            }
+
+            // handle empty appends
+            if (retainableBytes == null)
+                return true;
+            long length = retainableBytes.remaining();
+            if (length == 0)
+                return true;
+
+            // If we are already aggregating, and the content will fit, and the pass buffer is mostly empty then just aggregate
+            if (_aggregate != null && _aggregate.space() >= length && (length * 100) < retainableBytes.maxSize())
+                return _aggregate.append(retainableBytes.getByteBuffer());
+
+            // If the content is a tiny part of the retainable, then better to aggregate rather than accumulate
+            if (shouldAggregate(retainableBytes, length))
+                return append(retainableBytes.getByteBuffer());
+
+            // We will accumulate, so stop any further aggregation without allocating a new aggregate buffer;
+            _aggregate = null;
+
+            // Do we have space?
+            long space = _maxSize - size();
+            if (length <= space)
+            {
+                // We have space, so add a retained slice;
+                _buffers.add(retainableBytes.slice());
+                retainableBytes.skip(length);
+                return true;
+            }
+
+            // Are we full?
+            if (space == 0)
+                return false;
+
+            // Add a space limited retained slice of the buffer
+            length = space;
+            _buffers.add(retainableBytes.slice(length));
+            retainableBytes.skip(length);
+            return false;
+        }
+
+        @Override
+        public Mutable add(ByteBuffer bytes) throws ReadOnlyBufferException, BufferOverflowException
+        {
+            if (LOG.isDebugEnabled())
+                LOG.debug("add BB {} <- {}", this, BufferUtil.toDetailString(bytes));
+            add(RetainableByteBuffer.wrap(bytes));
+            return this;
+        }
+
+        @Override
+        public Mutable add(RetainableByteBuffer bytes) throws ReadOnlyBufferException, BufferOverflowException
+        {
+            if (LOG.isDebugEnabled())
+                LOG.debug("add RBB {} <- {}", this, bytes);
+            long size = size();
+            long space = _maxSize - size;
+            long length = bytes.size();
+            if (space < length)
+                throw new BufferOverflowException();
+
+            if (shouldAggregate(bytes, length) && append(bytes))
+            {
+                bytes.release();
```

Comment:
This breaks the `Retainable` contract: it's up to the caller to release.

Reply:
No. See javadoc for add.  It explicitly says that responsibility is passed.  This is how add differs from append.